### DISCUSSION TITLE
Quarto and Emacs

### DISCUSSION THREAD

**Discussion by plantarum** [quarto-dev/quarto-cli#483] [Q&A] [answered]
Hi,

I've put a bit of work into my RMarkdown workflow in Emacs, and it works quite well. Now I'm wondering about switching to Quarto:

1. Are there any plans to support an Emacs workflow?
2. Is Quarto in a state where it would make sense to start updating packages to accommodate it?

I'm fairly familiar with the key parts of Emacs that I think would be needed (at least for R coding): ESS, markdown-mode, polymode. I would be happy to help test and extend these for use with Quarto. But that only makes sense if Quarto development has reached a reasonably stable state.

Thanks,

Tyler

**Comment by jjallaire** [comment, marked as answer]
Yes, in fact we already have a package on MELPA: https://github.com/quarto-dev/quarto-emacs

There is one major change to it that we are hoping to complete soon which will provide much better integration with Quarto render/preview. Aside from that we'd very much welcome any input or direct help you might able to provide!

cc @cscheid

**Comment by cscheid** [comment]
~~Hi there! I will note that just last night I noticed that there's something wrong with the M-x package-install workflow through melpa, and I'm sad to report that quarto-mode won't currently work there. (I'm tracking down an interaction with the underlying polymode library)~~

~~With that said, if you download and add [quarto-mode.el](https://github.com/quarto-dev/quarto-emacs) directly to your .elisp or .emacs.d directory, and load it "the old-fashioned way", then it should work fine. I'm sorry about the current inconvenience, and I'll fix this in the coming days.~~

`quarto-mode` is working on MELPA, go ahead and use it! 🎉

**Comment by jcmkk3** [comment]
This may be a bit of a diversion from this question. I'm happy to move it to a separate discussion, if that makes sense, but I've also been thinking about how quarto might fit into an emacs based workflow. I'm a bit of an editor geek and try to keep up with what is happening in the emacs ecosystem so maybe the below will be helpful.

There has been a push recently to start offering packages through [NonGNU Elpa](https://elpa.nongnu.org/nongnu/) instead/in addition to MELPA. NonGNU Elpa is a recent distribution channel with its main advantage being that it will be enabled by default in future emacs versions, but is not as restrictive about what can be included as GNU Elpa. It might be useful to keep this distribution channel in mind for the future.

[Transient](https://github.com/magit/transient) will be part of the upcoming emacs 28 release. It is most known as one of the primary interaction models behind the popular Magit package. There might be some nice use cases for using that sort of interaction model for a package like Quarto, which has a cli at its core. I could imagine that it could be nice for some of the export options where a user might want to set different option flags.

It feels like many popular recent packages have focused on being more composable and leveraging existing emacs functionality like completing-read for minibuffer completion systems ([vertico](https://github.com/minad/vertico), [selectrum](https://github.com/raxod502/selectrum), [mct](https://gitlab.com/protesilaos/mct), built-in fido/icomplete modes), completion-at-point for in-buffer completions ([corfu](https://github.com/minad/corfu), [company-capf](https://github.com/company-mode/company-mode/blob/master/company-capf.el), [eglot](https://github.com/joaotavora/eglot), [LSP Mode](https://emacs-lsp.github.io/lsp-mode/page/main-features/)), flymake for linting, and tempo for templating.

There have been some interesting new-ish packages around literate and interactive programming like [code-cells](https://github.com/astoff/code-cells.el), [comint-mime](https://github.com/astoff/comint-mime), [emacs-jupyter](https://github.com/nnicandro/emacs-jupyter), [nrepl](https://github.com/nrepl/nrepl), [monroe](https://github.com/sanel/monroe).

There have been some nice improvements to the visual look and feel of emacs with Protesilaos Stavrou's accessible [Modus themes](https://protesilaos.com/emacs/modus-themes) being part of emacs 28. Emoji support and variable width font support has also been improved. Nicolas Rougier has been pushing the limits of what is possible within emacs visually. Some of their experiments include [notebook-mode](https://github.com/rougier/notebook-mode). Daniel Mendler has also shown some of the visual posibilities of emacs with [org-modern](https://github.com/minad/org-modern) and [osm](https://github.com/minad/osm).

**Comment by cscheid** [comment]
@jcmkk3 I would love for the quarto emacs story to be well supported in modern emacs ecosystems. 

With that said, I'm _very much not an expert_ on them, and there are other projects within quarto for at least the next half year or so that will take most of my time. So I don't expect I personally will have time to learn all of this stuff. Still, if there are folks in the community who want to take the lead on a more modern quarto+emacs infrastructure, I should be able to help them. For example, you mentioned LSP mode, and developing a fully-fledged LSP for quarto is in our roadmap.

**Comment by harishnkr** [comment]
Is there any update to the package? It feels like many features are missing, so my question is how are people using emacs running quarto?

**Reply by mcanouil** [reply]
Please don't resurrect 3 years old thread which are clearly outdated or at least unlikely to be up to date.
Instead, open a new discussion and reference this one in it.

Also please go not the repository given in the reply marked as the answer.

- https://github.com/quarto-dev/quarto-emacs
